src/Simple/JSON.purs

```diff
@@ -76,47 +78,57 @@ read = readImpl
 class ReadForeign a where
   readImpl :: Foreign -> F a
 
-instance readForeign :: ReadForeign Foreign where
+instance readForeignForeign :: ReadForeign Foreign where
   readImpl = pure
 
-instance readChar :: ReadForeign Char where
+instance readForeignChar :: ReadForeign Char where
   readImpl = readChar
 
-instance readNumber :: ReadForeign Number where
+instance readForeignNumber :: ReadForeign Number where
   readImpl = readNumber
 
-instance readInt :: ReadForeign Int where
+instance readForeignInt :: ReadForeign Int where
   readImpl = readInt
 
-instance readString :: ReadForeign String where
+instance readForeignString :: ReadForeign String where
   readImpl = readString
 
-instance readBoolean :: ReadForeign Boolean where
+instance readForeignBoolean :: ReadForeign Boolean where
   readImpl = readBoolean
 
-instance readArray :: ReadForeign a => ReadForeign (Array a) where
+instance readForeignArray :: ReadForeign a => ReadForeign (Array a) where
   readImpl = readElements <=< readArray
     where
       readElements xs = sequence $ readImpl <$> xs
 
-instance readNullOrUndefined :: ReadForeign a => ReadForeign (NullOrUndefined a) where
+instance readForeignNullOrUndefined :: ReadForeign a => ReadForeign (NullOrUndefined a) where
   readImpl = readNullOrUndefined readImpl
 
-instance readMaybe :: ReadForeign a => ReadForeign (Maybe a) where
+instance readForeignMaybe :: ReadForeign a => ReadForeign (Maybe a) where
   readImpl = map unNullOrUndefined <<< readImpl
 
-instance readNullable :: ReadForeign a => ReadForeign (Nullable a) where
+instance readForeignNullable :: ReadForeign a => ReadForeign (Nullable a) where
   readImpl o = withExcept (map reformat) $
     map toNullable <$> traverse readImpl =<< readNull o
     where
       reformat error = case error of
         TypeMismatch inner other -> TypeMismatch ("Nullable " <> inner) other
         _ -> error
 
-instance readStrMap :: ReadForeign a => ReadForeign (StrMap.StrMap a) where
+instance readForeignStrMap :: ReadForeign a => ReadForeign (StrMap.StrMap a) where
   readImpl = sequence <<< StrMap.mapWithKey (const readImpl) <=< readStrMap
 
-instance readRecord ::
+instance readForeignTuple :: (ReadForeign a, ReadForeign b) => ReadForeign (Tuple a b) where
+  readImpl = asTuple <=< readArray
+    where asTuple :: Array Foreign -> F (Tuple a b)
+          asTuple = case _ of
+            [a, b] -> do
+              ra <- readImpl a
+              rb <- readImpl b
+              pure $ Tuple ra rb
+            l -> fail $ TypeMismatch "2 values" (show (length l) <> " values")
```

Comment:
Oh wait, I just realized now that you have a single array decoding. For reading, is there not a way we could somehow maybe read in the whole array and then map them back to the tuples? Might require some generics-rep code?

Reply:
Like (a /\ b /\ c /\ d) should be [a,b,c,d]?

Reply:
Sounds good. Separate PR?

Reply:
Realized that some actually do encode nested array pairs though. I think probably it'd be best to go with whatever you need with this for now.

Reply:
Oh, nested tuples don't serialise correctly, I'll try to fix it.

Reply:
n-tuples serialisation should really be a flat array as you proposed, even more considering that horrible `{}`.

Reply:
Shouldn't `Data.Foreign` provide `readUnit`?

Reply:
`unit` is really supposed to be an opaque value. That should be a empty record in the type definition.

Reply:
I'm not sure I understood what you mean... I need `ReadForeign Unit` for encoding Tuple3 & friends and used a broken `readImpl` for now, but I guess `Data.Foreign` could provide that.

Reply:
You should never be using the value of Unit anywhere for reading or writing, it's only passed around like an opaque `foreign import data` declared type. You should *always* use an empty record for this.

Reply:
Unit's foreign representation should be `null` or `undefined` or something anyway, the value is absolutely meaningless because it's never supposed to be inspected or used.

Reply:
Is that what you mean? How would that be better?
In any case, a generic n-tuple serialisation as array probably won't need unit anyway, right?

Reply:
Well, as in no one will give you a stop element anyway. People expect (1 /\ 2 /\ 3) to become [1,2,3]